### PR TITLE
[3.0] Add preferredProtocol fallback strategy

### LINKED ISSUE #8599: triple不支持应用级地址发现
- [ ] I have searched the [issues](https://github.com/apache/dubbo/issues) of this repository and believe that this is not a duplicate.
- [ ] I have checked the [FAQ](https://github.com/apache/dubbo/blob/master/FAQ.md) of this repository and believe that this is not a duplicate.

### Environment

* Dubbo version: 3.x
* Operating System version: Ubuntu 18.04.5
* Java version: 11
* nacos version: 1.4.2

### Steps to reproduce this issue

1. provider配置
```
dubbo:
  application:
    metadata-type: remote
    register-mode: instance
  registry:
    address: nacos://${nacos.host}:${nacos.port}
  protocol:
    name: tri
    serialization: protobuf
```
2. consumer配置
```
dubbo:
  application:
    metadata-type: remote
    service-discovery:
      migration: APPLICATION_FIRST
  registry:
    address: nacos://${nacos.host}:${nacos.port}
#    ?registry-type=${nacos.registry-type}
  protocol:
    name: tri
    serialization: protobuf
```

Pls. provide [GitHub address] to reproduce this issue.

### Expected Result

image

### Actual Result

image

If there is an exception, please attach the exception trace:

java.lang.IllegalStateException: Failed to check the status of the service service.HelloService. No provider available for the service service.HelloService from the url dubbo://

### LINKED ISSUE #8698: [3.0] The logic of preferredProtocol was not user-friendly, we maybe should optimize it.
### What's the problem?
When we only use the `triple` protocol in dubbo with setting provider's register-mode as `instance`, the consumer can't find the service.

The source is in `org.apache.dubbo.config.metadata.ServiceInstanceHostPortCustomizer`, let's have a look:

```java
@Override
public void customize(ServiceInstance serviceInstance) {
    String host = null;
    int port = -1;
    // ......
    Set<URL> urls = writableMetadataService.getExportedServiceURLs();
    if (CollectionUtils.isNotEmpty(urls)) {
        ApplicationModel applicationModel = serviceInstance.getApplicationModel();
        String preferredProtocol = applicationModel.getCurrentConfig().getProtocol();
        //  The default value of preferredProtocol is dubbo.
        if (preferredProtocol != null) {
            for (URL exportedURL : urls) {
                // The protocol of exportedURL is tri
                // because of tri != dubbo, the host and port will not be assigned values.
                if (preferredProtocol.equals(exportedURL.getProtocol())) {
                    host = exportedURL.getHost();
                    port = exportedURL.getPort();
                    break;
                }
            }
        } else {
            URL url = urls.iterator().next();
            host = url.getHost();
            port = url.getPort();
        }
        // If not the port of instace > 0, the application instance will not be registered to RegistryCenter(like Nacos, Zookeeper).
        // As a result, the consumer can't find the avaliabe providers when our settging is registry-mode=instance
        if (serviceInstance instanceof DefaultServiceInstance) {
            DefaultServiceInstance instance = (DefaultServiceInstance) serviceInstance;
            instance.setHost(host);
            instance.setPort(port);
        }
    }
}
```

For the moment, if you want both consumer and provider to force using `Application Discovery` in tri protocol, just change one line config of provider as follows:

```yaml
dubbo:
  application:
    metadata-type: remote
    register-mode: instance
    protocol: tri  <==== HERE, set preferredProtocol as tri
```

### The possible solution
change the default `application-protocol` as `TRIPLE` in the future
> Why not now? Because most users are still trying to migrate from 2.0 to 3.0

```java
// => org.apache.dubbo.config.ApplicationConfig#checkDefault
protected void checkDefault() {
    super.checkDefault();
    if (protocol == null) {
        protocol = DUBBO;   // <======= HERE
    }
    // ......
}
```
In `ServiceInstanceHostPortCustomizer`, If the default protocol not exists, just try to get the first protocol and log an error log.

Related to #8666 
Related to #8599

### PR DESCRIPTION
**Pull request by plusmancn**
Fixes #8698 
Fixes #8599
Related to #8666

### CODE REVIEW

File: dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/metadata/ServiceInstanceHostPortCustomizer.java [hunk 1]
```diff
@@ -56,11 +60,19 @@ public void customize(ServiceInstance serviceInstance) {
...
                 }
+                
+                if (host == null || port == -1) {
+                    logger.warn("The default preferredProtocol \"" + preferredProtocol + "\" is not found, fall back to the strategy that pick the first found protocol. Please try to modify the config of dubbo.application.protocol");
+                    URL url = urls.iterator().next();
+                    host = url.getHost();
+                    port = url.getPort();
+                }
             } else {
                 URL url = urls.iterator().next();
```
Comment (guohao) [inline dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/metadata/ServiceInstanceHostPortCustomizer.java:71]: It is undeterminable that which protocol will be selected

Comment (EarthChen) [reply dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/metadata/ServiceInstanceHostPortCustomizer.java:71]: Set does not guarantee order

**Review by guohao — APPROVED** [review]
LGTM

**Review by EarthChen — APPROVED** [review]
LGTM

**Comment by plusmancn** [comment]
Don't merge now, the unit test still has a little problem. I am trying to solve this.

File: dubbo-config/dubbo-config-api/src/test/java/org/apache/dubbo/config/metadata/ServiceInstanceHostPortCustomizerTest.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    
+    @BeforeAll
+    public static void setUp() {
+        serviceInstanceHostPortCustomizer = new ServiceInstanceHostPortCustomizer();
+    }
+    
+    @Test
+    void customizePreferredProtocol() {
+        ApplicationModel applicationModel = ApplicationModel.defaultModel();
+        applicationModel.getApplicationConfigManager().setApplication(new ApplicationConfig("service-preferredProtocol"));
```
[the remark this answers is not in the document]
Comment (plusmancn) [reply dubbo-config/dubbo-config-api/src/test/java/org/apache/dubbo/config/metadata/ServiceInstanceHostPortCustomizerTest.java:45]: Isolate `applicationModel` from default using multi-instacne feature

**Comment by plusmancn** [comment]
> Don't merge now, the unit test still has a little problem. I am trying to solve this.

Done! The problem of the unit test had been fixed.
It can be merged. @AlbumenJ

**Pull request merged**